### ISSUE TITLE
No contest menu

### ISSUE THREAD

**Issue by dgiffen** [CDrummond/lms-musicsimilarity#2]
I got the script going and all seems well. Finished with no errors and wrote the Jukebox. However, I get no context menus on any song for the plug in. Under the DSTM there is a new entry called Similar. When I select this, the script appears to add songs, but only 2 are added to the que.

**Comment by CDrummond** [comment]
Which skin? I have had another user state that they do not appear with the default skin, but they appear with Material Skin (which is all I use). Must admit I do not use these entries.

As to the number of tracks added, you can configure this in the plugin's settings. If you have set a higher value and still only 2 are added, then I _think_ the reason wil be that the python backened filtered out too many tracks. If you pass "-l DEBUG" when running the backend you will get more info as to why tracks are filtered.

**Comment by dgiffen** [comment]
thanks, I have not tried other skins. I did try the extGUILMS and they do not show up there either. I usually use Orange Squeeze on my phone, so it isn't a critical issue. Material Skin installed and they are indeed there. Thanks. I will fiddle with the settings. Right now it only selects a couple of tracks and the regular DSTM plug in takes over. I can tell because it adds Tidal tracks and this should not, if I understand this.

**Comment by CDrummond** [comment]
If no tracks are recevied from the backend then LastMix and/or Spotify will be used. This was to handle the case where no local tracks were in the queue.

All I can suggest, as stated before, is to run the backend with debug and see what is happening.

**Comment by dgiffen** [comment]
It is for sure working. Is there anyway to pipe the debug output to a text file or something more easily read than the terminal window. I selected Tom Petty, Free Falling. I have over 24k tracks and I see lots of ouput in the terminal, but it added 0 tracks to this mix as similar tracks. 0 Tracks for Similar Tracks for Artist. Then only 2  tracks for a MIX and one of them was another version of Free Falling. My collection is mostly stuff in the same vein. this is the last few lines of output
2022-01-04 08:15:33 D similar_tracks: 0, filtered_tracks::meta: 736, filtered_tracks::attribs: 20989
2022-01-04 08:15:33 D Add some tracks from filtered_tracks::meta, 0/736
2022-01-04 08:15:33 D Path:file:///media/storage/music/Main/Tom%20Petty%20And%20The%20Heartbreakers/Playback/D03-01%20Free%20Fallin%27.flac 0.029096
2022-01-04 08:15:33 D Path:file:///media/storage/music/Main/Tom%20Petty/Wildflowers/11%20To%20Find%20a%20Friend.flac 0.174843
2022-01-04 08:15:33 I 127.0.0.1 - - [04/Jan/2022 08:15:33] "POST /api/similar HTTP/1.0" 200 -

**Comment by CDrummond** [comment]
As you can see 736 tracks were filtered due to metadata - probably same artist, etc. 20989 were filtered due to attributes - diffferent key, bpm, etc.

If you are running on Linux:

```
./music-similarity.py -l DEBUG > out.log 2>&1
```

**Comment by dgiffen** [comment]
OK, Thanks. Could not for the life of me figure out how to pipe this to a log. Ubuntu 20, so worked perfect. I haven't really looked at it yet. It just doesn't seem right that out of 24K tracks and probably half my library would be considered to be similar to Tom Petty only 2 TRACKS are considered similar. I think its working, but it evidently is being far more precise than I would have thought. I also use Plexamp and it has similar using some sort of sonic analysis. Track mixes through it find many more tracks and they are at least to my ears within reason. Thanks for the help. I see a fair number of parameters that I could change and maybe that will produce something closer to what I was expecting. I am certain the issue isn't the plugin itself, just my settings need some fine tuning.

**Comment by CDrummond** [comment]
What the script does is:
1. Ask's musly for ~5000 tracks similar to seed track
2. For each of these, it then checks to see they are a different track (e.g. match on title), different artist (depending on setting), etc. These are the "filtered_tracks::meta" tracks.
3. It then checks the tracks using Essentia data - to see they are within a BPM range, have similar key, etc. These are the "filtered_tracks::attrib" tracks.

It is this Essentia checking that is filtering out too much. You could look at the music-similarity.db file, using any SQLite DB browser.

Perhaps you need to adjust the settings in LMS?

**Comment by dgiffen** [comment]
Thanks, I will try some more stuff. I would just think here that out of 24K tracks, I would have more than 2 that sound similar. I have tried at least 20 main stream FM radio stuff and it never produces more than 2 tracks. With the default settings of the plugin. I can see for sure it looks through a  fair number of candidates, but it gets really choosy at the end. From a strictly similar artist concept, nearly half of my collection would revolve around Springsteen, Petty, Seger, etc. So again logically I would guess a mix of 50 would be reasonable. I will drop this as to the script itself is working for sure, fine tuning is expected

**Comment by CDrummond** [comment]
Yes, I'm 100% sure you should have more than 2. What is actually happening is that ALL tracks are being filtered out. At the end the script checks there are at least 2. There are not, so it takes two from the filtered-by-meta list.

Please, you need to see why the filtering is happening. It's obvious something is wrong, but I don't have access to your system, library, or settings. Hence why you should look at the results in the DB, check why the script says its filtering out tracks, etc.

**Comment by dgiffen** [comment]
I am looking at the data now. I think something hosed the data. I did have a brief power outage, but when I started the script up, it appeared to start somewhere in the middle, which sounded right. the data row with file, artist, title etc is all just in the file and nothing but NULL everywhere else. Except for the artists with Z. also explains why almost all the tracks it picks are artists with Z. I am going to move a few hundred files to another folder and see if that doesn't sort this out. I am fairly confident though that redoing this will cure the issues.

**Comment by CDrummond** [comment]
You _should_ be able to re-run the script and it will re-analyse where necessary.

**Comment by dgiffen** [comment]
I think something corrupted the database writing. The Z artiists have all sections of the data row filled out. Including Artist, Track Album etc. All others have the file filled out and then null on the Artist Track Album. Even though it counts 24K tracks only 50 or so of them actually have all the data correct. So I am fairly certain its the entire database that needs redone. 24K tracks will take at least 24 hours, so I am going to make a temp smaller set and see I am correct. I will for sure update you on this.

**Comment by CDrummond** [comment]
If you do re-run the analysis, set the "paths.cache" to some folder (see docs/OtherConfig.md). This will then save the Essentia analysis to JSON files. If you subsequently need to re-analyse then these will be read instead. No such cache for Musly as all data stored in DB - and Musly s, for me, very fast.

**Comment by dgiffen** [comment]
I did rerun the analysis. All is working perfectly now. I am getting mixes more like I expected. Looking at the database is what solved this, thank you so much for that tip. Thanks again for making this. I tried in vain to get musly and essentia going in the past, but this made it simple.

**Comment by CDrummond** [comment]
I'll leave this open for now, as it is valid - the entries are not shown in the default web UI. I looked into copying this from the LMS MusicMagick plugin - but is just way too much code/effort for a UI and feature that I do not use. So, won't be spending time to resolve this.
